File: src/lib/repositories/settingsRepository.ts
import { run, getAll } from '@/lib/db';

export interface SettingRow {
    key: string;
    value: string;
}

export async function getSetting(key: string): Promise<string | undefined> {
    const row = await (await import('@/lib/db')).getFirst<SettingRow>(
        `SELECT value FROM settings WHERE key = ?`, [key]
    );
    return row?.value ?? undefined;
}

export async function setSetting(key: string, value: string): Promise<void> {
    await run(
        `INSERT INTO settings (key, value, updated_at) VALUES (?, ?, ?)
         ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at`,
        [key, value, Date.now()]
    );
}

export async function deleteSetting(key: string): Promise<void> {
    await run(`DELETE FROM settings WHERE key = ?`, [key]);
}

export async function getAllSettings(): Promise<Record<string, string>> {
    const rows = await getAll<SettingRow>(`SELECT * FROM settings`);
    const map: Record<string, string> = {};
    for (const r of rows) map[r.key] = r.value;
    return map;
}

export async function deleteAllSettings(): Promise<void> {
    await run(`DELETE FROM settings`);
}

File: mda_flutter/lib/data/database/repositories/settings_repository.dart
/// Settings repository — the de-facto preference store (SPEC §7).
/// All app preferences live in the `settings` table as key/value pairs.
library;

import '../db.dart';

class SettingsRepository {
  Future<String?> getSetting(String key) async {
    final row = await getFirst('SELECT value FROM settings WHERE key = ?', [key]);
    final value = row?['value'];
    return value is String ? value : null;
  }

  /// Upsert with a fresh updated_at timestamp.
  Future<void> setSetting(String key, String value) async {
    await run(
      '''INSERT INTO settings (key, value, updated_at) VALUES (?, ?, ?)
         ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at''',
      [key, value, DateTime.now().millisecondsSinceEpoch],
    );
  }

  Future<void> deleteSetting(String key) async {
    await run('DELETE FROM settings WHERE key = ?', [key]);
  }

  Future<Map<String, String>> getAllSettings() async {
    final rows = await getAll('SELECT * FROM settings');
    return {
      for (final row in rows)
        if (row['key'] is String && row['value'] is String)
          row['key'] as String: row['value'] as String,
    };
  }

  Future<void> deleteAllSettings() async {
    await run('DELETE FROM settings');
  }
}
